xbmc/utils/ExecString.cpp

```diff
@@ -40,7 +40,8 @@ CExecString::CExecString(const std::string& function,
 {
   m_valid = !m_function.empty() && !target.GetPath().empty();
 
-  m_params.emplace_back(StringUtils::Paramify(target.GetPath()));
+  m_params.emplace_back(StringUtils::Paramify(
+      target.IsVideoDb() && target.IsVideoAssetNav() ? target.GetDynPath() : target.GetPath()));
```

Comment:
I don't think this is the right place to fix the problem.

Reply:
Thinking about it, placing here has the widest impact, even for non-playback actions so I should probably test for the type of action, or move a little closer to the context menu in the call stack.
I'll leave a todo that the better fix is to support extras or any asset in the videodb:// path.

Based on previous explanations of what path/dynpath are, converting the path should be correct though? We're only interested in the filename in order to play, and not in the path?
I'll also do some testing with non-local files.

Reply:
path == identity of the entity, should never be changed during lifetime of the instance (although we're still doing this in our code base)

dynpath == the path (URL) to be used for playback of a media item - can change dynamically during instance lifetime, thus the name

Reply:
I copied the above from the path that pops up the dialog after selecting a movie in the library.
This makes a copy of the info of the item, the item remains unchanged, there is no SetPath.

After digging, the playback of versions from context menu is also broken for multiple versions of same type for a movie, with the same reason.

Selecting in library is fine for any version and extra because it makes a copy of the item and applies the same trick I copied here.

I added two commits/different approaches to fix: 1/ better condition in CExecString to avoid side-effects 2/ move the path shenanigans into the caller (cannot be higher than that), which is almost a copy/paste of the code that makes versions/extras work from a selection in the library view (see CGUIWindowVideoBase::OnPlayMedia())

Let me know why one you prefer... I know neither one is really satisfying. The alternative is to remove the 4 context menu buttons (all broken in slightly different ways)...

Reply:
If I need to opt for one of them, clearly 2/ - it uses path trickery we already do elsewhere.

Reply:
Looking at 2/ - I do not understand why this is actually needed. Look at the implementation of `SetPathAndPlay`: it already does the trick with setting the path, just before instantiating the `CVideoSelectActionprocessor`.  There should be no need to correct the path one more time inside `CVideoSelectActionprocessor`implementation.'

Or is it that `if (!item->m_bIsFolder && item->IsVideoDb())` evaluates not to true, because of this damn design that the item can be a movie and a folder at the same time? Imo this line would be where the root cause needs to be fixed.

Reply:
Addendum: Imo an extra never should have a videoed URL as path, because extras are not contained in the video db.

Reply:
Maybe the fix is as simple as replacing

`if (!item->m_bIsFolder && item->IsVideoDb())`with
`if ((!item->m_bIsFolder || item->HasVideoVersions() || item->HasVideoExtras) && item->IsVideoDb())`?

Reply:
SetPathAndPlay() is not hit from library selection or context menu. You may be on to something here.

Reply:
After some testing:
- SetPathAndPlay is used in Play, Play using, Play version using, Play extra using, not in Choose version / extra
- not sure how SetPathAndPlay could help when the ChooseDialog is used to make a choice. The altered item is not the one returned by the dialog and that gets processed. I think the only case the original item is returned is when there is a single version and the dialog is skipped.
- still, the Play version using and Play extra using with SetPathAndPlay work 100% > it's a coincidence. Works due to different working of the external player handler
- yes when the movie is a folder, the code you highlighted is bypassed. That doesn't prevent the Play XXX using... from working correctly, reinforcing the idea that external player behaves differently from video player
- SetPathAndPlay is not in the code flow for Choose version or Choose extra

So adding the path override or item copy+path override after a selection by the choose dialog seems reasonable?

Adding in CVideoPlayActionProcessorBase::ProcessAction() right after CVideoVersionHelper::ChooseVideoFromAssets() is possible but I'm not sure because the settings action can be Info and no a form of play. It doesn't seem to break the info dialog though. This is coded in the last commit, approach 3/.

Reply:
Imo this needs to be rethought carefully. I'm not really overseeing the consequences of changes 1/ to 3/... but I know that you are touching were sensitive central code areas where all kind of items are processed... We need to be really careful and must really understand what we are changing.

As I said, multiple things seem to come together here:

1) extras must never have a videoed url as path, because they are not part of the video db
2) items representing a default movie version and a folder at the same time are problematic (e.g. not handled correctly In `SetPathAndPlay`
3) two items pointing to different movies must not have the same path. Currently, we allow to have multiple movies with the same version (id) as children of a "parent" movie which leads to items pointing to different movies having the same video db url as path. That's something that must be fixed!